### PR TITLE
refactor(core): refactor `CloudFormationLang.toJSON()`

### LINKED ISSUE #13465: (core): Templating error on custom-resource introduced with version 1.92
````typescript
for (let index = 0; index < existingVpc.availabilityZones.length; index++) {
      let getEndpointIp = new AwsCustomResource(this, idPrefix + 'GetEndpointIp' + index, {
        policy: AwsCustomResourcePolicy.fromSdkCalls({resources: AwsCustomResourcePolicy.ANY_RESOURCE}),
        onUpdate: {
          action: 'describeNetworkInterfaces',
          service: 'EC2',
          parameters: {NetworkInterfaceIds: vpcendpoint.vpcEndpointNetworkInterfaceIds},
          physicalResourceId: PhysicalResourceId.of('DirectoryId'),
          outputPath: 'NetworkInterfaces.' + index + '.PrivateIpAddress'
        }
      })
      targetGroup.addTarget(new IpTarget(Token.asString(getEndpointIp.getResponseField('NetworkInterfaces.' + index + '.PrivateIpAddress'))));
    }
````

We are doing this, because we need to put an api gateway behind a load balancer (company policy). If there is a better solution to do this, feedback is welcome.

### Reproduction Steps

this is the output with aws-cdk/core 1.91

````yaml
...
  announceGetEndpointIp0D5B4E67A:
    Type: Custom::AWS
    Properties:
      ServiceToken:
        Fn::GetAtt:
          - AWS12345678
          - Arn
      Create:
        action: describeNetworkInterfaces
        service: EC2
        parameters:
          NetworkInterfaceIds:
            Fn::GetAtt:
              - announcevpcannouncevpcendpoint55991BB5
              - NetworkInterfaceIds
        physicalResourceId:
          id: DirectoryId
        outputPath: NetworkInterfaces.0.PrivateIpAddress
      Update:
        action: describeNetworkInterfaces
        service: EC2
        parameters:
          NetworkInterfaceIds:
            Fn::GetAtt:
              - announcevpcannouncevpcendpoint55991BB5
              - NetworkInterfaceIds
        physicalResourceId:
          id: DirectoryId
        outputPath: NetworkInterfaces.0.PrivateIpAddress
      InstallLatestAwsSdk: true
    DependsOn:
      - announceGetEndpointIp0CustomResourcePolicyC5EFFE47
    UpdateReplacePolicy: Delete
    DeletionPolicy: Delete
    Metadata:
      aws:cdk:path: announce-dev-local/announce-GetEndpointIp0/Resource/Default
...
````

this is the output with aws-cdk/core 1.92

```yaml
...
  announceGetEndpointIp0D5B4E67A:
    Type: Custom::AWS
    Properties:
      ServiceToken:
        Fn::GetAtt:
          - AWS12345678
          - Arn
      Create:
        Fn::Join:
          - ""
          - - '{"action":"describeNetworkInterfaces","service":"EC2","parameters":{"NetworkInterfaceIds":"'
            - Fn::GetAtt:
                - announcevpcannouncevpcendpoint55991BB5
                - NetworkInterfaceIds
            - '"},"physicalResourceId":{"id":"DirectoryId"},"outputPath":"NetworkInterfaces.0.PrivateIpAddress"}'
      Update:
        Fn::Join:
          - ""
          - - '{"action":"describeNetworkInterfaces","service":"EC2","parameters":{"NetworkInterfaceIds":"'
            - Fn::GetAtt:
                - announcevpcannouncevpcendpoint55991BB5
                - NetworkInterfaceIds
            - '"},"physicalResourceId":{"id":"DirectoryId"},"outputPath":"NetworkInterfaces.0.PrivateIpAddress"}'
      InstallLatestAwsSdk: true
    DependsOn:
      - announceGetEndpointIp0CustomResourcePolicyC5EFFE47
    UpdateReplacePolicy: Delete
    DeletionPolicy: Delete
    Metadata:
      aws:cdk:path: announce-dev-local/announce-GetEndpointIp0/Resource/Default
...
```

### What did you expect to happen?

After using `cdk deploy` we expected the same behavior as in version 1.91 with successful deployment.

### What actually happened?

The error only occurs during a `cdk deploy` with version 1.92.

```sh
26/44 | 8:32:42 AM | CREATE_FAILED        | Custom::AWS                               | announce-GetEndpointIp2/Resource/Default (announceGetEndpointIp20B447BE7) Template error: every Fn::Join object requires two parameters, (1) a string delimiter and (2) a list of strings to be joined or a function that returns a list of strings (such as Fn::GetAZs) to be joined.
	new CustomResource (/builds/marceldarvis/toolbox/services/announce-pipeline-modules-service/node_modules/@aws-cdk/core/lib/custom-resource.ts:119:21)
	\_ new AwsCustomResource (/builds/marceldarvis/toolbox/services/announce-pipeline-modules-service/node_modules/@aws-cdk/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts:374:27)
```

### Environment

  - **CDK CLI Version  :1.92**
  - **Framework Version:None**
  - **Node.js Version:v14.16.0** 
  - **OS               :linux**
  - **Language (Version):TypeScript (4.2.3)** 

### Other

We assume that this issue is related to this change:

- https://github.com/aws/aws-cdk/commit/3cb31043a42b035f6dcd2a318836d4bfc4973151

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by rix0rrr**
Our previous implementation of `toJSON()` was quite hacky.

It replaced values inside the structure with objects that had a custom
`toJSON()` serializer, and then called `JSON.stringify()` on the result.

The resulting JSON would have special markers in it where the Token
values would be string-substituted back in.

It's actually easier and gives us more control to just implement
JSONification ourselves in a Token-aware recursive function.

This change has been split off from a larger, upcoming PR in order
to make the individual reviews smaller.

Incidentally also fixes #13465, as the type of encoded tokens is assumed to match
the type of the encoded value (e.g., a `string[]`-encoded token is assumed to
produce a list at deploy-time and so will not be quoted).

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts [hunk 1]
```diff
@@ -97,44 +47,213 @@ export class CloudFormationLang {
...
+ * 3. Otherwise, the type of the value it resolves to (string, number, complex object, ...)
+ *    determines how the value is rendered.
  */
-function deepQuoteStringsForJSON(x: any): any {
-  if (typeof x === 'string') {
-    // Whenever we escape a string we strip off the outermost quotes
-    // since we're already in a quoted context.
-    const stringified = JSON.stringify(x);
-    return stringified.substring(1, stringified.length - 1);
+function tokenAwareStringify(root: any, space: number, ctx: IResolveContext) {
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:104]: Add @param documentation

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts [hunk 2]
```diff
@@ -204,3 +323,59 @@ export function isNameOfCloudFormationIntrinsic(name: string): boolean {
...
+    if (value !== undefined) {
+      yield [key, value];
+    }
+  }
+}
+
+/**
+ * Quote string literals inside an intrinsic
+ */
+function quoteInsideIntrinsic(x: any): any {
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:352]: This feels awfully fragile. Any way to avoid having to encode a heuristic for each fn?

File: packages/@aws-cdk/core/test/cloudformation-json.test.ts [hunk 3]
```diff
@@ -1,12 +1,32 @@
...
-  'string tokens can be JSONified and JSONification can be reversed'(test: Test) {
-    const stack = new Stack();
+let app: App;
+let stack: Stack;
+beforeEach(() => {
+  app = new App();
+  stack = new Stack(app, 'Stack');
+});
+
+test('JSONification of literals looks like JSON.stringify', () => {
```
Comment (eladb) [inline packages/@aws-cdk/core/test/cloudformation-json.test.ts:12]: I am wondering if we can pull out an existing test suite for JSON.stringify to ensure we cover the spec correctly...

**Review by eladb — APPROVED** [review]

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts (hunk 2 above)
Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:352]: Not really. We can encode it in a different place and make it more official, but fundamentally this is what we have to do.

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts [hunk 4]
```diff
@@ -97,44 +47,227 @@ export class CloudFormationLang {
...
+      if (isIntrinsic(obj) && resolvedTypeHint(obj)) {
+        return renderIntrinsic(obj);
+      }
+
+      return renderCollection('{', '}', definedEntries(obj), ([key, value]) => {
+        if (key.startsWith(INTRINSIC_KEY_PREFIX)) {
+          [key, value] = value;
+        }
+
+        recurse(key);
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:181]: Is there a case where `key` is not a string? Should we be defensive against it?

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts [hunk 5]
```diff
@@ -97,44 +47,227 @@ export class CloudFormationLang {
...
+      throw new Error('This shouldnt happen anymore');
+    }
+    if (Array.isArray(obj)) {
+      return renderCollection('[', ']', obj, recurse);
+    }
+    if (typeof obj === 'object' && obj != null && !(obj instanceof Date)) {
+      // Treat as an intrinsic if this LOOKS like a CFN intrinsic (`{ Ref: ... }`)
+      // AND it's the result of a token resolution. Otherwise, we just treat this
+      // value as a regular old JSON object (that happens to look a lot like an intrinsic).
+      if (isIntrinsic(obj) && resolvedTypeHint(obj)) {
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:172]: doesn't `isIntrinsic()` impliy `resolvedTypeHint()` always returns true? Maybe just assert?

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts [hunk 6]
```diff
@@ -204,3 +337,64 @@ export function isNameOfCloudFormationIntrinsic(name: string): boolean {
...
+function* definedEntries<A extends object>(xs: A): IterableIterator<[string, any]> {
+  for (const [key, value] of Object.entries(xs)) {
+    if (value !== undefined) {
+      yield [key, value];
+    }
+  }
+}
+
+/**
+ * Quote string literals inside an intrinsic
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:364]: I still feel I don't understand why this is needed. Maybe some examples? Why isn't it possible to just use `JSON.stringify()` of the intrinsic ?

File: packages/@aws-cdk/core/test/cloudformation-json.test.ts [hunk 7]
```diff
@@ -232,22 +292,47 @@ nodeunitShim({
         },
       },
     });
+  });
+
+  test('Intrinsics can occur in key position', () => {
```
Comment (eladb) [inline packages/@aws-cdk/core/test/cloudformation-json.test.ts:297]: Very cool!

**Review by eladb — COMMENTED** [review]
A few questions

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts (hunk 4 above)
Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:181]: `Object.entries()` will always and only return strings.

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts (hunk 5 above)
Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:172]: We can have type hints on things that aren't intrinsics (for example, as returned by a lazy).

We can also have structures that LOOK like intrinsics but shouldn't be treated as such (for example, the literal structure `{ Ref: X }`)

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts (hunk 6 above)
Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:364]: I will add more realistic tests.

Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:364]: Oh here is one: https://github.com/aws/aws-cdk/blob/huijbers/refactor-tojson/packages/@aws-cdk/core/test/cloudformation-json.test.ts/#L183

Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:364]: Your intuition was correct though. I missed some edge cases. After thinking about it some more, just a straight up recursion over all strings should also do it. Added more tests to confirm.

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts (hunk 5 above)
Comment (eladb) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:172]: > We can have type hints on things that aren't intrinsics (for example, as returned by a lazy).

This conditional won't cover non-intrinsics (it's an `&&`).

> We can also have structures that LOOK like intrinsics but shouldn't be treated as such (for example, the literal structure { Ref: X })
 
Shouldn't `isIntrinsic()` cover this?

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts (hunk 6 above)
Comment (eladb) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:364]: 👍
